flink-runtime/src/test/java/org/apache/flink/runtime/util/profiler/ProfilingServiceTest.java

```diff
@@ -0,0 +1,170 @@
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+/** Unit tests for {@link ProfilingService}. */
+public class ProfilingServiceTest {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ProfilingServiceTest.class);
+    private static final Configuration configs = new Configuration();
+    private static final String NO_ACCESS_TO_PERF_EVENTS = "No access to perf events.";
+    private static final String NO_ALLOC_SYMBOL_FOUND = "No AllocTracer symbols found.";
+    private static final String resourceID = "TestJobManager";
+    private static final long profilingDuration = 3L;
+    private static final int historySizeLimit = 3;
+
+    private ProfilingService profilingService;
+
+    @BeforeAll
+    static void beforeAll() {
+        configs.set(RestOptions.MAX_PROFILING_HISTORY_SIZE, historySizeLimit);
+    }
+
+    @BeforeEach
+    void setUp(@TempDir Path tempDir) {
+        configs.set(RestOptions.PROFILING_RESULT_DIR, tempDir.toString());
+        profilingService = ProfilingService.getInstance(configs);
+        verifyConfigsWorks(profilingService, tempDir);
+    }
+
+    @AfterEach
+    void tearDown() throws IOException {
+        profilingService.close();
+    }
+
+    @Test
+    public void testSingleInstance() throws IOException {
+        ProfilingService instance = ProfilingService.getInstance(configs);
+        Assertions.assertEquals(profilingService, instance);
+        instance.close();
+    }
+
+    @Test
+    void testFailedRequestUnderProfiling() throws ExecutionException, InterruptedException {
+        ProfilingInfo profilingInfo =
+                profilingService
+                        .requestProfiling(resourceID, 10, ProfilingInfo.ProfilingMode.ITIMER)
+                        .get();
+        Assertions.assertEquals(ProfilingInfo.ProfilingStatus.RUNNING, profilingInfo.getStatus());
+        try {
+            profilingService
+                    .requestProfiling(
+                            resourceID, profilingDuration, ProfilingInfo.ProfilingMode.ITIMER)
+                    .get();
+            Assertions.fail("Duplicate profiling request should throw with IllegalStateException.");
+        } catch (Exception e) {
+            Assertions.assertTrue(e.getCause() instanceof IllegalStateException);
+        }
+    }
+
+    @Test
+    @Timeout(value = 1, unit = TimeUnit.MINUTES)
+    public void testAllProfilingMode() throws ExecutionException, InterruptedException {
+        for (ProfilingInfo.ProfilingMode mode : ProfilingInfo.ProfilingMode.values()) {
+            ProfilingInfo profilingInfo =
+                    profilingService.requestProfiling(resourceID, profilingDuration, mode).get();
+            if (isNoPermissionOrAllocateSymbol(profilingInfo)) {
+                LOG.warn(
+                        "Ignoring failed profiling instance in {} mode, which caused by no permission.",
```

Comment:
I think there exist more messages other than `no permission` here.

Reply:
Yes, It will print the detailed message in the revised version.